x-pack/plugins/cases/public/components/actions/tags/edit_tags_selectable.tsx

```diff
@@ -0,0 +1,398 @@
+
+      dispatch({ type: Actions.CHECK_TAG, payload: selectedTags });
+      dispatch({ type: Actions.UNCHECK_TAG, payload: unSelectedTags });
+      onChangeTags({ selectedTags, unSelectedTags });
+    },
+    [onChangeTags, state.tags]
+  );
+
+  const onNewItem = useCallback(
+    (newTag: string) => {
+      const { selectedTags, unSelectedTags } = getSelectedAndUnselectedTags(options, state.tags);
+      dispatch({ type: Actions.CHECK_TAG, payload: [newTag] });
+      setSearchValue('');
+      onChangeTags({ selectedTags: [...selectedTags, newTag], unSelectedTags });
+    },
+    [onChangeTags, options, state.tags]
+  );
+
+  const onSelectAll = useCallback(() => {
+    dispatch({ type: Actions.CHECK_TAG, payload: Object.keys(state.tags) });
+    onChangeTags({ selectedTags: Object.keys(state.tags), unSelectedTags: [] });
+  }, [onChangeTags, state.tags]);
+
+  const onSelectNone = useCallback(() => {
+    const unSelectedTags = [];
+
+    for (const [label, tag] of Object.entries(state.tags)) {
+      if (tag.tagState === TagState.CHECKED || tag.tagState === TagState.PARTIAL) {
+        unSelectedTags.push(label);
+      }
+    }
+
+    dispatch({ type: Actions.UNCHECK_TAG, payload: unSelectedTags });
+    onChangeTags({ selectedTags: [], unSelectedTags });
+  }, [state.tags, onChangeTags]);
+
+  /**
+   * TODO: Remove hack when PR https://github.com/elastic/eui/pull/6317
+   * is merged and the new fix is merged into Kibana.
+   *
+   * This is a hack to force a rerender when
+   * the user adds a new tag. There is a bug in
+   * the EuiSelectable where a race condition that's causing the search bar
+   * to not to match terms with the empty string to trigger the reload.
+   * This means that when a user press the button to add a tag the
+   * search bar clears but the options are not shown.
+   */
+  const [_, setRerender] = useState(0);
+
+  useEffect(() => {
+    if (isEmpty(searchValue)) {
+      setRerender((x) => x + 1);
+    }
+  }, [options, setRerender, searchValue]);
+
+  return (
+    <EuiSelectable
+      options={options}
+      searchable
+      searchProps={{
+        placeholder: i18n.SEARCH_PLACEHOLDER,
+        isLoading,
+        isClearable: !isLoading,
+        onChange: setSearchValue,
+        value: searchValue,
+      }}
+      renderOption={renderOption}
+      listProps={{ showIcons: false }}
+      onChange={onChange}
+      noMatchesMessage={<NoMatchesMessage searchValue={searchValue ?? ''} onNewItem={onNewItem} />}
+      data-test-subj="cases-actions-tags-edit-selectable"
+    >
+      {(list, search) => (
+        <>
+          {search}
+          <EuiSpacer size="s" />
+          <EuiFlexGroup
+            alignItems="center"
+            justifyContent="spaceBetween"
+            responsive={false}
+            direction="row"
+          >
+            <EuiFlexItem>
+              <EuiText size="xs" color="subdued">
+                {i18n.TOTAL_TAGS(tags.length)}
```

Comment:
Should we update the total as a user is selecting/unselecting tags in the flyout? Basically should it reflect the currently selected tags? I guess that gets tricky with partially selected tags? Maybe initially it would include partially selected ones.

Reply:
I think this counter shows all available tags. @mdefazio Is this assumption correct?

Reply:
My vote is that the counter would show the number of tags matching / partially matching the text entered into the search box, as I think this is confusing:

<img width="230" alt="image" src="https://user-images.githubusercontent.com/7405507/200379771-c46b3b58-2bb4-42f3-b147-47b5d94310c1.png">

Or if it's going to show the total number of tags, then it should display that in the label, such as `Total tags available: 19`.

Reply:
Great point @peteharverson.  I'd vote for `Total tags available: 19` to keep things straightforward initially.

Reply:
What do you think of

<img width="624" alt="Screenshot 2022-11-08 at 11 07 48 AM" src="https://user-images.githubusercontent.com/7871006/200559652-4e5d87e9-a2ef-42f5-bbc0-a3032e49dd52.png">

I slightly change the text to match what we have in the cases table when you select cases and I added the "Selected tags section"

Reply:
So this is a bit nitpicky, but I prefer having the number in the same spot for both phrases.  And I think it might be better to include 'Total' so it's not mistaken that this may change based on the search bar. Lastly, I don't think we need to repeat 'Tags' in both. 

I appreciate wanting to matching the Cases table, but I'm not sure that's the best example to follow and should likely update that to match what we do here... 🙃   

`18 total tags  |  9 selected` _or_ ` Total tags: 18  |  Selected: 9`

Reply:
Done! 
<img width="433" alt="Screenshot 2022-11-08 at 2 17 15 PM" src="https://user-images.githubusercontent.com/7871006/200561834-4c8f8193-2275-465d-9b70-715b94c2a207.png">